providers/dns/internal/clientdebug/client.go

```diff
@@ -0,0 +1,52 @@
+package clientdebug
+
+import (
+	"net/http"
+	"net/http/httputil"
+	"os"
+	"strconv"
+)
+
+type DumpTransport struct {
+	rt http.RoundTripper
+}
+
+func NewDumpTransport(rt http.RoundTripper) *DumpTransport {
+	if rt == nil {
+		rt = http.DefaultTransport
+	}
+
+	return &DumpTransport{rt: rt}
+}
+
+func (d *DumpTransport) RoundTrip(h *http.Request) (*http.Response, error) {
+	dump, _ := httputil.DumpRequestOut(h, true)
+
+	println("[HTTP Request]")
+	println(string(dump))
```

Comment:
Avoid `print` and `println`. From the spec:

> These functions are documented for completeness but are not guaranteed to stay in the language.
> 
> https://go.dev/ref/spec#Bootstrapping

```suggestion
	fmt.Println("[HTTP Request]")
	fmt.Println(string(dump))
```